src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc
+{
+    using System;
+    using System.Buffers;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.Serialization;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft;
+    using Nerdbank.Streams;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+    using StreamJsonRpc.Protocol;
+
+    /// <summary>
+    /// Uses Newtonsoft.Json serialization to serialize <see cref="JsonRpcMessage"/> as JSON (text).
+    /// </summary>
+    public class JsonMessageFormatter : IJsonRpcMessageTextFormatter
+    {
+        /// <summary>
+        /// The key into an <see cref="Exception.Data"/> dictionary whose value may be a <see cref="JToken"/> that failed deserialization.
+        /// </summary>
+        internal const string ExceptionDataKey = "JToken";
```

Comment:
not that I think it matters much, but I suppose you could use nameof(JToken) here

Reply:
I agree it doesn't really matter since the term will never be renamed. But I don't think it's appropriate here because the string isn't actually locked to that identifier. It's part of a public *protocol* contract for looking up the error. If the identifier were named something differently later, we would still keep this string to avoid breaking the wire protocol. It's just that the wire protocol was inspired by the name of a thing we're using.

---
In reply to: [218242767](https://github.com/Microsoft/vs-streamjsonrpc/pull/160#discussion_r218242767) [](ancestors = 218242767)